client/asset/eth/nodeclient.go

```diff
@@ -531,6 +547,215 @@ func (n *nodeClient) sendSignedTransaction(ctx context.Context, tx *types.Transa
+	txOpts, _ := n.txOpts(ctx, 0, 3e5, dexeth.GweiToWei(maxFeeRate))
+	if err := n.addSignerToOpts(txOpts); err != nil {
+		return nil, fmt.Errorf("addSignerToOpts error: %w", err)
+	}
+
+	boundTokenContract := bind.NewBoundContract(tokenAddr, *dexerc20.ERC20_ABI, n.ec, n.ec, n.ec)
+
+	return boundTokenContract.Transact(txOpts, "approve", n.tokenSwapContractAddr, amount)
+}
+
+// initiateToken initiates multiple token swaps in the same transaction.
+func (n *nodeClient) initiateToken(ctx context.Context, initiations []v0.ERC20SwapInitiation, token common.Address, maxFeeRate uint64) (tx *types.Transaction, err error) {
+	// TODO: reject if there is duplicate secret hash
+	// TODO: use estimated gas
+	txOpts, _ := n.txOpts(ctx, 0, 1e6, dexeth.GweiToWei(maxFeeRate))
+	if err := n.addSignerToOpts(txOpts); err != nil {
+		return nil, fmt.Errorf("addSignerToOpts error: %w", err)
+	}
+
+	return n.tokenSwapContract.Initiate(txOpts, initiations, token)
+}
+
+// redeemToken redeems multiple token swaps. All redemptions must involve the
+// same ERC20 token.
+func (n *nodeClient) redeemToken(ctx context.Context, redemptions []v0.ERC20SwapRedemption, maxFeeRate uint64) (tx *types.Transaction, err error) {
+	// TODO: reject if there is duplicate secret hash
+	// TODO: use estimated gas
+	txOpts, _ := n.txOpts(ctx, 0, 300000, dexeth.GweiToWei(maxFeeRate))
+	if err := n.addSignerToOpts(txOpts); err != nil {
+		return nil, fmt.Errorf("addSignerToOpts error: %w", err)
+	}
+
+	return n.tokenSwapContract.Redeem(txOpts, redemptions)
+}
+
+// tokenIsRedeemable checks if a token swap identified by secretHash is currently
+// redeemable using secret.
+func (n *nodeClient) tokenIsRedeemable(ctx context.Context, secretHash, secret [32]byte) (bool, error) {
+	callOpts := &bind.CallOpts{
+		Pending: true,
+		From:    n.address(),
+		Context: ctx,
+	}
+	return n.tokenSwapContract.IsRedeemable(callOpts, secretHash, secret)
+}
+
+// refundToken refunds a token swap.
+func (n *nodeClient) refundToken(ctx context.Context, secretHash [32]byte, maxFeeRate uint64) (tx *types.Transaction, err error) {
+	// TODO: use estimated gas
+	txOpts, _ := n.txOpts(ctx, 0, 5e5, dexeth.GweiToWei(maxFeeRate))
+	if err := n.addSignerToOpts(txOpts); err != nil {
+		return nil, fmt.Errorf("addSignerToOpts error: %w", err)
+	}
+
+	return n.tokenSwapContract.Refund(txOpts, secretHash)
+}
+
+// tokenIsRefundable checks if a token swap identified by secretHash is
+// refundable.
+func (n *nodeClient) tokenIsRefundable(ctx context.Context, secretHash [32]byte) (bool, error) {
+	callOpts := &bind.CallOpts{
+		Pending: true,
+		From:    n.address(),
+		Context: ctx,
+	}
+	return n.tokenSwapContract.IsRefundable(callOpts, secretHash)
+}
+
+// estimateInitTokenGas checks the amount of gas that is used to redeem
+// token swaps.
+func (n *nodeClient) estimateInitTokenGas(ctx context.Context, tokenAddr common.Address, num int) (uint64, error) {
```

Comment:
Unlike with ETH, the gas usage will vary depending on each token's implementation of `transfer` and `transferFrom`.